File: packages/backend/spi-simulator/src/iso20022/schemas.ts
export interface GroupHeader {
  MsgId: string
  CreDtTm: string
  NbOfTxs: string
  TtlIntrBkSttlmAmt: string
  IntrBkSttlmDt: string
  SttlmInf: {
    SttlmMtd: string
  }
}

export interface CreditTransferTransaction {
  PmtId: {
    EndToEndId: string
    TxId?: string
  }
  InstgAgt: {
    FinInstnId: {
      ClrSysMmbId: {
        MmbId: string
      }
    }
  }
  Dbtr: {
    Nm?: string
  }
  DbtrAcct?: {
    Id: {
      Iban?: string
      Othr?: {
        Id: string
      }
    }
  }
  DbtrAgt: {
    FinInstnId: {
      ClrSysMmbId: {
        MmbId: string
      }
    }
  }
  CdtrAgt: {
    FinInstnId: {
      ClrSysMmbId: {
        MmbId: string
      }
    }
  }
  Cdtr: {
    Nm?: string
  }
  CdtrAcct?: {
    Id: {
      Iban?: string
      Othr?: {
        Id: string
      }
    }
  }
  IntrBkSttlmAmt: string
  ChrgBr: string
  CdtTrfTxInf?: CreditTransferTransaction
}

export interface Pacs008Document {
  FIToFICstmrCdtTrf: {
    GrpHdr: GroupHeader
    CdtTrfTxInf: CreditTransferTransaction | CreditTransferTransaction[]
  }
}

export interface OriginalGroupInformation {
  OrgnlMsgId: string
  OrgnlMsgNmId: string
  OrgnlCreDtTm: string
}

export interface OriginalTransactionReference {
  IntrBkSttlmAmt: string
  SttlmInf: {
    SttlmMtd: string
  }
  CdtrAgt: {
    FinInstnId: {
      ClrSysMmbId: {
        MmbId: string
      }
    }
  }
  DbtrAgt: {
    FinInstnId: {
      ClrSysMmbId: {
        MmbId: string
      }
    }
  }
  Cdtr: {
    Nm?: string
  }
  Dbtr: {
    Nm?: string
  }
  CdtTrfTxInf?: CreditTransferTransaction
}

export interface TransactionInformation {
  OrgnlInstrId?: string
  OrgnlEndToEndId: string
  OrgnlTxId?: string
  TxSts: string
  StsRsnInf?: {
    Rsn: {
      Cd: string
    }
    AddtlInf?: string
  }
  OrgnlGrpInf?: OriginalGroupInformation
  OrgnlTxRef?: OriginalTransactionReference
}

export interface Pacs002Document {
  FIToFIPmtStsRpt: {
    GrpHdr: GroupHeader
    TxInfAndSts: TransactionInformation | TransactionInformation[]
  }
}

export interface ResolutionOfInvestigation {
  OrgnlGrpInf: OriginalGroupInformation
  OrgnlPmtInfAndSts: {
    OrgnlEndToEndId: string
    OrgnlTxId?: string
    TxSts: string
    StsRsnInf?: {
      Rsn: {
        Cd: string
      }
      AddtlInf?: string
    }
  }
}

export interface PaymentReturn {
  PmtId: {
    EndToEndId: string
    TxId?: string
  }
  IntrBkSttlmAmt: string
  ChrgBr: string
  InstgAgt: {
    FinInstnId: {
      ClrSysMmbId: {
        MmbId: string
      }
    }
  }
  DbtrAgt: {
    FinInstnId: {
      ClrSysMmbId: {
        MmbId: string
      }
    }
  }
  CdtrAgt: {
    FinInstnId: {
      ClrSysMmbId: {
        MmbId: string
      }
    }
  }
  Dbtr: {
    Nm?: string
  }
  Cdtr: {
    Nm?: string
  }
  RtrInf: {
    Rsn: {
      Cd: string
    }
    AddtlInf?: string
  }
  OrgnlGrpInf: OriginalGroupInformation
  OrgnlPmtInfAndSts: {
    OrgnlEndToEndId: string
    OrgnlTxId?: string
    TxSts: string
  }
}

export interface Pacs004Document {
  FIToFIPmtRtr: {
    GrpHdr: GroupHeader
    TxInf: PaymentReturn | PaymentReturn[]
  }
}

File: packages/backend/spi-simulator-go/internal/iso20022/schemas.go
package iso20022

import "encoding/xml"

// Pacs008 - Credit Transfer (FIToFICstmrCdtTrf)
type Pacs008Document struct {
	XMLName                    xml.Name                     `xml:"Document"`
	FIToFICstmrCdtTrf         FIToFICstmrCdtTrf            `xml:"FIToFICstmrCdtTrf"`
}

type FIToFICstmrCdtTrf struct {
	XMLName                xml.Name                   `xml:"FIToFICstmrCdtTrf"`
	GrpHdr                 GroupHeader                `xml:"GrpHdr"`
	CdtTrfTxInf            []CreditTransferTransaction `xml:"CdtTrfTxInf"`
}

type GroupHeader struct {
	MsgId    string `xml:"MsgId"`
	CreDtTm  string `xml:"CreDtTm"`
	NbOfTxs  int    `xml:"NbOfTxs"`
	SttlmInf struct {
		SttlmMtd string `xml:"SttlmMtd"`
	} `xml:"SttlmInf"`
}

type CreditTransferTransaction struct {
	PmtId struct {
		EndToEndId string `xml:"EndToEndId"`
		TxId       string `xml:"TxId"`
	} `xml:"PmtId"`
	InstgAgt struct {
		FinInstnId struct {
			ClrSysId struct {
				MmbId string `xml:"MmbId"`
			} `xml:"ClrSysId"`
		} `xml:"FinInstnId"`
	} `xml:"InstgAgt"`
	IntrBkSttlmAmt struct {
		Value float64 `xml:",chardata"`
		Ccy   string  `xml:"Ccy,attr"`
	} `xml:"IntrBkSttlmAmt"`
	ChrgBr string `xml:"ChrgBr"`
	Dbtr struct {
		Nm string `xml:"Nm"`
	} `xml:"Dbtr"`
	DbtrAcct struct {
		Id struct {
			IBAN string `xml:"IBAN"`
		} `xml:"Id"`
	} `xml:"DbtrAcct"`
	DbtrAgt struct {
		FinInstnId struct {
			ClrSysId struct {
				MmbId string `xml:"MmbId"`
			} `xml:"ClrSysId"`
		} `xml:"FinInstnId"`
	} `xml:"DbtrAgt"`
	CdtrAgt struct {
		FinInstnId struct {
			ClrSysId struct {
				MmbId string `xml:"MmbId"`
			} `xml:"ClrSysId"`
		} `xml:"FinInstnId"`
	} `xml:"CdtrAgt"`
	Cdtr struct {
		Nm string `xml:"Nm"`
	} `xml:"Cdtr"`
	CdtrAcct struct {
		Id struct {
			IBAN string `xml:"IBAN"`
		} `xml:"Id"`
	} `xml:"CdtrAcct"`
}

// Pacs002 - Status Report (FIToFIPmtStsRpt)
type Pacs002Document struct {
	XMLName            xml.Name              `xml:"Document"`
	FIToFIPmtStsRpt    FIToFIPmtStsRpt      `xml:"FIToFIPmtStsRpt"`
}

type FIToFIPmtStsRpt struct {
	GrpHdr              GroupHeader              `xml:"GrpHdr"`
	TransactionInformationAndStatus []TransactionInformation `xml:"TxInfAndSts"`
}

type TransactionInformation struct {
	OrgnlEndToEndId string `xml:"OrgnlEndToEndId"`
	OrgnlTxId       string `xml:"OrgnlTxId"`
	TxSts           string `xml:"TxSts"`
	StsRsnInf       *struct {
		Rsn struct {
			Cd string `xml:"Cd"`
		} `xml:"Rsn"`
		AddtlInf string `xml:"AddtlInf"`
	} `xml:"StsRsnInf,omitempty"`
}

// Pacs004 - Payment Return (FIToFIPmtRtr)
type Pacs004Document struct {
	XMLName         xml.Name           `xml:"Document"`
	FIToFIPmtRtr    FIToFIPmtRtr       `xml:"FIToFIPmtRtr"`
}

type FIToFIPmtRtr struct {
	GrpHdr           GroupHeader        `xml:"GrpHdr"`
	PmtRtr           []PaymentReturn    `xml:"PmtRtr"`
}

type PaymentReturn struct {
	PmtId struct {
		EndToEndId string `xml:"EndToEndId"`
		TxId       string `xml:"TxId"`
	} `xml:"PmtId"`
	IntrBkSttlmAmt struct {
		Value float64 `xml:",chardata"`
		Ccy   string  `xml:"Ccy,attr"`
	} `xml:"IntrBkSttlmAmt"`
	RtrInf struct {
		Rsn struct {
			Cd string `xml:"Cd"`
		} `xml:"Rsn"`
		AddtlInf string `xml:"AddtlInf,omitempty"`
	} `xml:"RtrInf"`
	OrgnlGrpInf struct {
		OrgnlMsgId string `xml:"OrgnlMsgId"`
	} `xml:"OrgnlGrpInf"`
	OrgnlPmtInfAndSts struct {
		OrgnlEndToEndId string `xml:"OrgnlEndToEndId"`
	} `xml:"OrgnlPmtInfAndSts"`
}
